File: 11_sieve_of_eratosthenes/count_non_divisible.cpp
#include <vector>

using namespace std;

vector<int> elt_counter(vector<int> &a) {
  int n = a.size();
  vector<int> ret(2 * n + 1, 0);
  for (int i = 0; i < n; i++) {
    ret[a[i]]++;
  }
  return ret;
}

vector<int> solution(vector<int> &a) {
  int n = a.size();
  vector<int> elt_cnts = elt_counter(a);
  vector<int> non_divs(2 * n + 1, n);
  vector<int> ret(n, 0);
  for (int i = 1; i < 2 * n + 1; i++) {
    if (elt_cnts[i] != 0) {
      for (int j = 1; i * j < 2 * n + 1; j++) {
        if (elt_cnts[i * j] == 0) {
          non_divs[i * j] -= elt_cnts[i];
        }
      }
    }
  }
  for (int i = 0; i < n; i++) {
    ret[i] = non_divs[a[i]];
  }
  return ret;
}

File: 11_sieve_of_eratosthenes/count_non_divisible.py
def elt_counter(a):
  n = len(a)
  ret = [0] * (2 * n + 1)
  for i in a:
    ret[i] += 1
  return ret

def solution(a):
  n = len(a)
  elt_cnts = elt_counter(a)
  non_divs = [n] * (2 * n + 1)
  ret = []

  for i in range(2 * n + 1):
    if elt_cnts[i] != 0:
      j = 1
      while (i * j < 2 * n + 1):
        if elt_cnts[i * j] != 0:
          non_divs[i * j] -= elt_cnts[i]
        j += 1

  for i in range(n):
    ret += [non_divs[a[i]]]

  return ret
